coderd/keyrotate/rotate_internal_test.go

```diff
@@ -0,0 +1,526 @@
+	})
+
+	t.Run("DoesNotRotateValidKeys", func(t *testing.T) {
+		t.Parallel()
+
+		var (
+			db, _       = dbtestutil.NewDB(t)
+			clock       = quartz.NewMock(t)
+			keyDuration = time.Hour * 24 * 7
+			logger      = slogtest.Make(t, nil).Leveled(slog.LevelDebug)
+			ctx         = testutil.Context(t, testutil.WaitShort)
+		)
+
+		kr := &rotator{
+			db:          db,
+			keyDuration: keyDuration,
+			clock:       clock,
+			logger:      logger,
+			features: []database.CryptoKeyFeature{
+				database.CryptoKeyFeatureWorkspaceApps,
+			},
+		}
+
+		now := dbnow(clock)
+
+		// Seed the database with an existing key
+		existingKey := dbgen.CryptoKey(t, db, database.CryptoKey{
+			Feature:  database.CryptoKeyFeatureWorkspaceApps,
+			StartsAt: now,
+			Sequence: 123,
+		})
+
+		// Advance the clock by 6 days, 23 hours. Once we
+		// breach the last hour we will insert a new key.
+		clock.Advance(keyDuration - 2*time.Hour)
+
+		err := kr.rotateKeys(ctx)
+		require.NoError(t, err)
+
+		keys, err := db.GetCryptoKeys(ctx)
+		require.NoError(t, err)
+		require.Len(t, keys, 1)
+		require.Equal(t, existingKey, keys[0])
+
+		// Advance it again to just before the key is scheduled to be rotated for sanity purposes.
+		clock.Advance(time.Hour - time.Second)
+
+		err = kr.rotateKeys(ctx)
+		require.NoError(t, err)
+
+		// Verify that the existing key is still the only key in the database
+		keys, err = db.GetCryptoKeys(ctx)
+		require.NoError(t, err)
+		require.Len(t, keys, 1)
+		requireKey(t, keys[0], existingKey.Feature, existingKey.StartsAt.UTC(), time.Time{}, existingKey.Sequence)
+	})
+
+	// Simulate a situation where the database was manually altered such that we only have a key that is scheduled to be deleted and assert we insert a new key.
+	t.Run("DeletesExpiredKeys", func(t *testing.T) {
+		t.Parallel()
+
+		var (
+			db, _       = dbtestutil.NewDB(t)
+			clock       = quartz.NewMock(t)
+			keyDuration = time.Hour * 24 * 7
+			logger      = slogtest.Make(t, nil).Leveled(slog.LevelDebug)
+			ctx         = testutil.Context(t, testutil.WaitShort)
+		)
+
+		kr := &rotator{
+			db:          db,
+			keyDuration: keyDuration,
+			clock:       clock,
+			logger:      logger,
+			features: []database.CryptoKeyFeature{
+				database.CryptoKeyFeatureWorkspaceApps,
+			},
+		}
+
+		now := dbnow(clock)
+
+		// Seed the database with an existing key
+		deletingKey := dbgen.CryptoKey(t, db, database.CryptoKey{
+			Feature:  database.CryptoKeyFeatureWorkspaceApps,
+			StartsAt: now.Add(-keyDuration),
+			Sequence: 789,
+			DeletesAt: sql.NullTime{
+				Time:  now,
+				Valid: true,
+			},
+		})
+
+		err := kr.rotateKeys(ctx)
+		require.NoError(t, err)
+
+		// We should only get one key since the old key
+		// should be deleted.
+		keys, err := db.GetCryptoKeys(ctx)
+		require.NoError(t, err)
+		require.Len(t, keys, 1)
+		requireKey(t, keys[0], deletingKey.Feature, deletingKey.DeletesAt.Time.UTC(), time.Time{}, deletingKey.Sequence+1)
```

Comment:
Not a problem with this test, but I think it illustrates the point that we probably _do_ need to implement querying for  key IDs on demand.  Here we're inserting a key with startsAt == oldkey.DeletesAt == now, so some Coderds will start using the key right away, but others might query just before insertion and not have the key.

This also will be a problem on startup with a brand new system: since rotation is going to be decoupled from the component that fetches the keys for signing/verifying, there is no guarantee there are _any_ keys available to you at the time you initially query, since initial "rotation" might still be in progress.

Reply:
>  I think it illustrates the point that we probably do need to implement querying for key IDs on demand.

Agreed especially for workspace proxies.

> This also will be a problem on startup with a brand new system: since rotation is going to be decoupled from the component that fetches the keys for signing/verifying, there is no guarantee there are any keys available to you at the time you initially query, since initial "rotation" might still be in progress.

Running `keyrotate.Start` blocks until an initial rotation is done but yeah if you're running HA and you fail to acquire the lock there's definitely window where you could start serving requests and the initial rotation is still ongoing 🤔 . 
If we use a blocking lock instead this problem should go away.